### PR TITLE
Angular 11 🔥

### PR DESCRIPTION
**Pull request by qmonmert**
Commands :
``` 
ng update
ng update @angular/cli @angular/core
``` 

---

Please make sure the below checklist is followed for Pull Requests.

-   [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [x] Tests are added where necessary
-   [x] [jhipster-online](https://github.com/jhipster/jhipster-online) was updated if necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (bellow reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

File: generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs
Comment (kaidohallik) [inline generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs:56]: Do we have code which forces us to use `relativeLinkResolution: 'legacy'`?

If yes then can we changes those to use Angular 11 default `relativeLinkResolution: 'corrected'`?

Comment (qmonmert) [reply generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs:56]: I don't know, it's this command:`ng update @angular/cli @angular/core` to migrate which updates this file

Comment (kaidohallik) [reply generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs:56]: I think we should migrate from `'legacy'` to `'corrected'`.  
Tests are failing because `DatePipe` has stricter type checking in Angular 11 and code needs to be changed.  
@qmonmert are you planning to work on this PR to get CI green?

Comment (qmonmert) [reply generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs:56]: yes I'm working on it tomorrow, I have already started (but not push on this PR)

File: generators/client/templates/angular/src/main/webapp/app/admin/metrics/metrics-modal-threads/metrics-modal-threads.component.html.ejs
Comment (kaidohallik) [inline generators/client/templates/angular/src/main/webapp/app/admin/metrics/metrics-modal-threads/metrics-modal-threads.component.html.ejs:53]: Can you rename variable `entry` to `thread`, because now this is not `Object` entry but `Thread` object.

File: generators/client/templates/angular/src/main/webapp/app/admin/metrics/metrics-modal-threads/metrics-modal-threads.component.html.ejs
Comment (kaidohallik) [inline generators/client/templates/angular/src/main/webapp/app/admin/metrics/metrics-modal-threads/metrics-modal-threads.component.html.ejs:20]: Currently we have 4 space spacing in HTML files. We should follow the same style.  
If we want to go for 2 space spacing in HTML files then this should be done systematically for all HTML files.

**Review by kaidohallik — CHANGES_REQUESTED** [review]
Great unnecessary pipe removal!  
I have couple of comments.

File: generators/client/templates/angular/src/main/webapp/app/admin/metrics/metrics-modal-threads/metrics-modal-threads.component.html.ejs
Comment (qmonmert) [reply generators/client/templates/angular/src/main/webapp/app/admin/metrics/metrics-modal-threads/metrics-modal-threads.component.html.ejs]: fix

File: generators/client/templates/angular/src/main/webapp/app/admin/metrics/metrics-modal-threads/metrics-modal-threads.component.html.ejs
Comment (qmonmert) [reply generators/client/templates/angular/src/main/webapp/app/admin/metrics/metrics-modal-threads/metrics-modal-threads.component.html.ejs]: done

File: generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs
Comment (qmonmert) [reply generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs:56]: @kaidohallik do you have an idea to fix `DatePipe` with `Dayjs` ?

Comment (kaidohallik) [reply generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs]: `DayJs`  has https://day.js.org/docs/en/display/format, maybe use this instead of `DatePipe`? If using methods in HTML then it's also good idea to switch to `changeDetection: ChangeDetectionStrategy.OnPush`, to avoid bad performance.

File: generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs
Comment (ctamisier) [inline generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs:56]: the default is 'corrected' in angular 11, should we just take it out in this case ?

Comment (kaidohallik) [reply generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs:56]: We need to check, does we have code which relies on `relativeLinkResolution`, if yes then migrate that code and then remove this.

File: generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs
Comment (ctamisier) [inline generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs]: squash this commit with the first one of this PR, as relativeLinkResolution has been added and then removed ?

Comment (qmonmert) [reply generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs]: done

File: generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs
Comment (wmarques) [reply generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs]: The pipes are pure by default so for me the OnPush strategy is not mandatory, the pipe code will be executed only when the input changes even if the Component is not using the OnPush strategy

Comment (kaidohallik) [reply generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs]: Pipes are replaced by methods, so `OnPush` is needed for good performance.

Comment (qmonmert) [reply generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs]: when I use `changeDetection: ChangeDetectionStrategy.OnPush` on my component, the list of entities does not display

Comment (kaidohallik) [reply generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs]: On data load use `this.changeDetector.markForCheck();` as this is done in `metrics.component.ts`.

Comment (qmonmert) [reply generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs]: yes it works

Comment (wmarques) [reply generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs]: Maybe we can use custom date pipes using dayjs instead of the Angular default ones ? This way we can remove the OnPush and provide our users some dayJS pipes. 
For me formatting an input for the display is why pipes are made for., WDYT ?

That can be done on a separate PR

Comment (kaidohallik) [reply generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs]: For `DayJs` pipes is library available: https://github.com/ljmerza/ngx-dayjs - so if you think we should go for pipes then first need discuss, use this library or not.

Comment (wmarques) [reply generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs]: The lib doesn't look maintained so I won't go for it, I would vote for putting two custom pipes in our shared folder

Comment (kaidohallik) [reply generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs]: Agree with custom pipes in our shared folder, there is already `duration.pipe` for `DayJs`. So I suggest to create `shared/dayjs` subfolder, move `duration.pipe` there and add additional pipes.

Comment (qmonmert) [reply generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs]: ok let's do this

Comment (wmarques) [reply generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs]: Maybe `shared/dates` WDYT ? 😄  For me that should be an abstract thing so we can easily change our date lib in the future

Comment (kaidohallik) [reply generators/client/templates/angular/src/main/webapp/app/app-routing.module.ts.ejs]: Yes, this looks better, but maybe not in plural, but `shared/date` like other folders are named?

File: generators/client/templates/angular/src/main/webapp/app/shared/date/format-medium-date.pipe.ts.ejs
Comment (wmarques) [inline generators/client/templates/angular/src/main/webapp/app/shared/date/format-medium-date.pipe.ts.ejs:1]: Can you add few unit tests to check that the date is correctly formatted ? That will help us if we switch lib like we did recently

Same goes for the other pipe 😄

**Review by wmarques — CHANGES_REQUESTED** [review]

**Comment by ctamisier** [comment]
Should these new formats be used in user-management.component.html.ejs ?

**Comment by qmonmert** [comment]
@ctamisier I think there is a bug on the main branch with `user-management.component.html.ejs` because dates are not displayed

File: generators/client/templates/angular/src/main/webapp/app/shared/date/format-medium-date.pipe.ts.ejs
Comment (qmonmert) [reply generators/client/templates/angular/src/main/webapp/app/shared/date/format-medium-date.pipe.ts.ejs:1]: done

**Review by kaidohallik — COMMENTED** [review]
Currently dates and datetimes in entity detail view are broken, already before this Angular 11 PR, but can you use those new pipes also there to fix entity detail view?

File: generators/client/files-angular.js
Comment (qmonmert) [reply generators/client/files-angular.js]: done

**Comment by qmonmert** [comment]
> Currently dates and datetimes in entity detail view are broken, already before this Angular 11 PR, but can you use those new pipes also there to fix entity detail view?

done

**Comment by qmonmert** [comment]
> Should these new formats be used in user-management.component.html.ejs ?

@ctamisier I don't think in this PR because `user.createdDate` is a `Date` and not a `dayjs.Dayjs`

**Comment by qmonmert** [comment]
@pascalgrimaud why there is one cancelled checks ? 🤔

**Comment by pascalgrimaud** [comment]
some timeout...

**Comment by qmonmert** [comment]
you can restart the test ?

**Comment by mshima** [comment]
It happens sometimes with that test. Restarted.

**Comment by qmonmert** [comment]
thanks @mshima

**Review by kaidohallik — APPROVED** [review]
Now everything LGTM, thanks @qmonmert

**Comment by qmonmert** [comment]
tests are ✅

**Comment by qmonmert** [comment]
@pascalgrimaud I claim my money 😄 https://opencollective.com/generator-jhipster/expenses/28243

**Comment by pascalgrimaud** [comment]
thanks for your work @qmonmert ! approved

**Comment by mshima** [comment]
We should bump the dependency `@angular-builders/custom-webpack@11.0.0-beta.1`

**Comment by wmarques** [comment]
@mshima id rather wait for a stable release

**Comment by mshima** [comment]
@wmarques `@angular-builders/custom-webpack@10` has dependency and imports angular@10.
So we are using a mixed of angular-cli 11 with 10.

I would rather wait for the entire stack to update.
Since we updated to angular 11, IMO a beta supporting 11 is better then a stable supporting 10.

**Comment by wmarques** [comment]
I do not agree here, the builder only add a "wrapper" around the CLI which will be in 11, so for me having the custom-webpack builder in 10 doesn't bother me a lot. 
I don't feel confortable by giving our user a beta dependency, I think the stable 11 will come pretty soon so we should better wait for it.

However I agree that the most simple solution was to wait for the release of the custom-webpack v11 😢

**Comment by MathieuAA** [comment]
We should not use dependencies in beta versions, unless we know for sure there won't be any issues or the risks are measured and taken into account.
If we can wait for the dependency to use angular-cli 11, then let's do it.
The only warning I raise is about our release planning: if we wanna do a beta release, then we should think about whether this matter would be an issue for our release planning.

Ideally, this issue should have been brought up and discussed before merging this PR. These kinds of issues should be taken into account the next time we do something like this.

**Comment by kaidohallik** [comment]
FYI: I made tests with HMR (`ng serve --hmr`) and:
* it works fully after upgrading `@angular-builders` components to `v11.0.0-beta.1`
* after upgrading `@angular-builders` components to `v11.0.0-beta.1` can remove `module.hot` stuff from `app.main.ts`, HMR works still fully
* HMR seems not working or working partially if not upgrading `@angular-builders` to `v11.0.0-beta.1` (our `module.hot` stuff in `app.main.ts` seems to be too simplified for full HMR support)
* in previous versions (Angular 10 and before) HMR seems not working or working partially

So upgrading Angular to v11 was great step, now everybody can switch manually `@angular-builders` dependencies to version `11.0.0-beta.1` and enjoy full HMR support.

**Pull request merged**
